### ISSUE TITLE
Could not push image: An image does not exist locally with the tag: spamhater

### ISSUE THREAD

**Issue by SpamhaterAlex** [bmuschko/gradle-docker-plugin#580]
Hi, 

I can build images, but not push. 

```
//define variables 
def dockerRegistryCredentials = new DockerRegistryCredentials()
dockerRegistryCredentials.email = rootProject.hasProperty('email') ? rootProject['email'] : 'default_value'
dockerRegistryCredentials.password = rootProject.hasProperty('password') ? rootProject['password'] : 'default_value'
dockerRegistryCredentials.username = rootProject.hasProperty('username') ? rootProject['username'] : 'default_value'
dockerRegistryCredentials.url = rootProject.hasProperty('url') ? rootProject['url'] : 'default_value'

//create Dockerfile to destination target-directory 
task createDockerfile(type: Dockerfile, dependsOn: copyFile)  {
    destFile = project.file('build/distributions/uberJar/Dockerfile')
    from 'openjdk:8'
}

//builds an image from a Dockerfile 
task buildImage(type: DockerBuildImage, dependsOn: createDockerfile) {
    inputDir = createDockerfile.destFile.parentFile
	tag = 'testapp'
        registryCredentials = dockerRegistryCredentials
}

//Pushes an image to a registry.
task pushImage(type: DockerPushImage, dependsOn: buildImage) {
    tag = 'testapp'
    imageName = 'testapp'
    registryCredentials = dockerRegistryCredentials
}
```

So, when I run this command: 
`gradle pushImage -Pemail=spamhater@gmail.com -Pusername=spamhater -Ppassword=xxxx -Purl=hub.docker.codemania.com `

I get this problem: 
```
---> Using cache
---> 5a5957da4edc
Successfully built 5a5957da4edc
Successfully tagged testapp:latest

> Task :buildImage
Created image with ID '5a5957da4edc'.

> Task :pushImage FAILED
Pushing image with name 'testapp:testapp'.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':pushImage'.
> Could not push image: An image does not exist locally with the tag: spamhater
```

normally, my repo in docker hub looks like this: 
`hub.docker.codemania.com/spamhater/bigapp:latest`

Would be the url right like in the console? 
I would be happy, if you can give me an hint, thanks in advice

**Comment by mailfly** [comment]
hi， this question is fixed ?
i encountered the same problem ,which one could be help me?

**Comment by orzeh** [comment]
@mailfly referring to the above example, in build image task `tag` should  be `testapp:testapp` (`[image name]:[tag]`)

**Comment by orzeh** [comment]
@mailfly your `buildImage` task will produce image with name like `my-fancy-project-0.0.1-snapshot.jar`. Then in `pushImage` you try to push an imgage with the same name but with tag `0.0.1-snapshot` (`project.version` value), so Docker is expecting that image `my-fancy-project-0.0.1-snapshot.jar:0.0.1-snapshot` is available, which isn't true I guess.

What I would do is:
````groovy
task buildImage(type: DockerBuildImage) {
	dependsOn createDockerfile
	inputDir = createDockerfile.destFile.parentFile
	tag = "${project.getName()}:${version}"
}
task pushImage(type: DockerPushImage) {
	dependsOn buildImage
	imageName = "${project.getName()}"
	tag = project.version
}
````

Hope it's now clear.

**Comment by mailfly** [comment]
@orzeh thanks for you reply , i try it , also have the  same issue :(

C:\source\code\fund-mock>gradle pushImage

> Configure project :
fund-mock
0.0.1-snapshot
Step 1/4 : FROM fabric8/java-centos-openjdk8-jdk:latest

---> de74f8152f3b
Step 2/4 : VOLUME ["/opt/app"]

---> Using cache
---> 79c3309f8d96
Step 3/4 : ENTRYPOINT ["java", "-jar", "/fund-mock.jar"]

---> Using cache
---> 22f368b2a838
Step 4/4 : LABEL maintainer=zoufeng02@ppdai.com

---> Using cache
---> ecde90be711b
Successfully built ecde90be711b
Successfully tagged fund-mock:0.0.1-snapshot

> Task :buildImage
Building image using context 'C:\source\code\fund-mock\build\docker'.
Using tag 'fund-mock:0.0.1-snapshot' for image.
Created image with ID 'ecde90be711b'.

> Task :pushImage FAILED
Pushing image with name 'fund-mock:0.0.1-snapshot'.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':pushImage'.
> Could not push image: An image does not exist locally with the tag: linsanfu

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 0s
3 actionable tasks: 2 executed, 1 up-to-date

**Comment by orzeh** [comment]
@mailfly `linsanfu` what is this? Obviously you have a problem with setting proper tag...

**Comment by mailfly** [comment]
@orzeh yes ,this proper tag i set is in below : it's same is a username value, how push the image will using this value? i'm confused , ：(

docker {
	url = 'tcp://10.X.X.X:2376'
	registryCredentials {
		//url = 'docker hub url ?'
		username = 'linsanfu' -- 
		password = 'XXX'
	}
}

**Comment by orzeh** [comment]
@mailfly please read the [Docker push](https://docs.docker.com/engine/reference/commandline/push/#extended-description) documentation to get a clarification how things work. Also look at our [functional test](https://github.com/bmuschko/gradle-docker-plugin/blob/master/src/functTest/groovy/com/bmuschko/gradle/docker/DockerWorkflowFunctionalTest.groovy#L237) for an example how to push to private repository. I hope it helps.

**Comment by mailfly** [comment]
@orzeh  thanks for you clarification， i will try , thanks again
